### ISSUE TITLE
Color Palettes theme support with a single array of colors argument

### ISSUE THREAD

**Issue by webmandesign** [WordPress/gutenberg#6425]
## Issue Overview

_Gutenberg version: **2.6.0**_

In my theme I am grabbing all theme mods colors into an array. I would like to use this array (only the unique values/colors from it) to set up color palettes for Gutenberg. But currently the color palettes theme support is declared by passing each color value as a separate argument, not a single array of colors.

**Could you please consider setting up color palettes with a single array argument instead of multiple ones?**

_(Example code below was taken from [Gutenberg theme](https://github.com/WordPress/gutenberg-starter-theme/blob/master/functions.php#L87).)_

## Current Behavior

```
add_theme_support( 'editor-color-palette',
	'#0073aa',
	'#229fd8',
	'#eee',
	'#444',
);
```

## Expected Behavior
This is just a rough, quick example to prove the point:
```
$colors = array_unique( array(
	get_theme_mod( 'color_accent', '#0073aa' ),
	get_theme_mod( 'color_accent_hover', '#229fd8' ),
	get_theme_mod( 'color_text', '#eee' ),
	get_theme_mod( 'color_headings', '#444' ),
) );
add_theme_support( 'editor-color-palette', $colors );
```
As you might imagine, theme user can simply decide to use the same color for `color_accent` and `color_accent_hover` and/or for `color_text` and `color_headings`. That's why I use `array_unique()` to get rid of duplicates. This wouldn't be possible if I just listed every color as a separate argument. Also, it seems to be simpler to work with array than with separate arguments when I want to automate things.

(Actually, currently I'm trying to figure out how to split my colors array into individual arguments and pass them to `add_theme_support` with no success... - forgive my ignorance if that's doable.)

Thank you for consideration on this!

Regards,

Oliver

**Comment by paulwilde** [comment]
It’s kinda weird how the current implementation doesn’t support passing in an array of arrays. Makes it quite annoying as you are required to add separate argument for each colour.

**Comment by webmandesign** [comment]
@samikeijonen Thanks Sami, I've actually seen this before I opened the issue. But the new approach you've linked to still does not resolve my original problem. It just creates multiple array arguments, while each array is for a single color. So, again, multiple arguments passed.

@paulwilde That's exactly my point here &rarr; I'd like to treat all colors in a single array argument.

**Comment by samikeijonen** [comment]
Sure, I was just pointing out if you missed that. I don't have an answer to your original question though. In other words I don't know the reason for it.

**Comment by chrisvanpatten** [comment]
```php
function g7g_format_colors( $color_map = array() ) {
	// Define some vars we'll use later
	$g7g_colors     = [];
	$g7g_colors_raw = [];

	// Loop through the color map array
	foreach ($color_map as $color_name => $color_default) {
		// Pull in the value
		$color = get_theme_mod($color_name, $color_default);

		// If the value is not in our "raw" color array, add it
		// (so it will be skipped next time) and add to the
		// actual array to pass to add_theme_support
		if ( ! in_array($g7g_colors_raw, $color) ) {
			$g7g_colors_raw[] = $color;

			$g7g_colors[] = array(
				'name'  => $color_name,
				'color' => $color,
			);
		}
	}

	return $g7g_colors;
}

// Array of color names and default values
$colors = array(
	'color_accent'       => '#DEFAULT',
	'color_accent_hover' => '#DEFAULT',
	'color_text'         => '#DEFAULT',
	'color_headings'     => '#DEFAULT',
);

add_theme_support( 'editor-color-palette', g7g_format_colors($colors) );
```

**Comment by chrisvanpatten** [comment]
Untested, but I think this would work:

```php
$colors = [
	[
		'name'  => 'name_here',
		'color' => '#HEXVAL',
	],
	[
		'name'  => 'name_here',
		'color' => '#HEXVAL',
	],
];

// Add 'editor-color-palette' as the first argument
$args = array_unshift($colors, 'editor-color-palette');

// Calls `add_theme_support` with arguments from the $args array
// (each array element is treated as an arg)
call_user_func_array('add_theme_support', $args);
```

I agree though, being able to pass in a single array would be nice!

**Comment by chrisvanpatten** [comment]
Here's a nicer version wrapped up in a function that also checks for duplicates:

```php
/**
 * Returns an array of color names and values in the multi-dimensional
 * array format expected by `add_theme_support`
 * 
 * @param $color_map array
 * @return array
 */
function g7g_register_editor_color_palette( $color_map = array() ) {
	// Define the main array of args
	$args = array(
		'editor-color-palette',
	);

	// Used to look for duplicate colors
	$colors = array();

	// Loop through the color map array
	foreach ( $color_map as $color_name => $color_default ) {
		// Pull in the value
		$color = get_theme_mod( $color_name, $color_default );

		// Only add color values we haven't already added
		if ( ! in_array( $colors, $color ) ) {
			// Add to $colors so it's skipped next time
			$colors[] = $color;

			// Each new color is a new argument to pass to
			// add_theme_support
			$args[] = array(
				'name'  => $color_name,
				'color' => $color,
			);
		}
	}

	call_user_func_array( 'add_theme_support', $args );
}

add_action( 'after_setup_theme', function () {
	// Array of color names and default values
	$colors = array(
		'color_accent'       => '#HEXVAL',
		'color_accent_hover' => '#HEXVAL',
		'color_text'         => '#HEXVAL',
		'color_headings'     => '#HEXVAL',
	);

	g7g_register_editor_color_palette($colors);
} );
```

**Comment by webmandesign** [comment]
@chrisvanpatten 

Thanks for [the code](#issuecomment-384362801). Indeed this would work with a minor fix - you shouldn't use [`array_unshift`](http://php.net/manual/en/function.array-unshift.php) to return value:
```
array_unshift( $colors, 'editor-color-palette' );
call_user_func_array( 'add_theme_support', $colors );
```

However, as you can see, this is not very pretty... In a theme setup function full of pretty `add_theme_support` code this one looks very weird, hacky. For example:
```
add_theme_support( 'title-tag' );
add_theme_support( 'custom-logo' );
add_theme_support( 'align-wide' );

array_unshift( get_theme_colors_array(), 'editor-color-palette' );
call_user_func_array( 'add_theme_support', $colors );
```
But surely, this seems to be the only way for me to go right now&hellip;

As for your [other code](#issuecomment-384364582), thanks for that too! I wasn't particularly trying to resolve the colors duplication issue with my original question, it was just a rough example of usage ;) But your code does prove to be very useful, thanks!

**Comment by webmandesign** [comment]
Hey guys, I've created a pull request #6437 fixing this issue. Works perfectly fine for me. Please test if you are interested.

**Comment by justintadlock** [comment]
Just ran into this issue trying to help out someone.  The second parameter should definitely accept an array of colors rather than adding each color as an additional parameter.  I was starting to think I was going to need to perform some black magic to get things working.

**Comment by webmandesign** [comment]
Please disregard the PR #6437 due to messy rebase. I've just created a new PR #7025 implementing the single array setup functionality.

**Comment by webmandesign** [comment]
New, updated fix (matching current master commits): #7619
